azurerm/resource_arm_operational_insight_workspace.go

```diff
@@ -0,0 +1,224 @@
+	//}
+	d.SetId(*workspace.ID)
+
+	return resourceArmOperationalInsightWorkspaceRead(d, meta)
+
+}
+
+func resourceArmOperationalInsightWorkspaceRead(d *schema.ResourceData, meta interface{}) error {
+	// I don't understand why we can get the meta data. How the framework set it.
+	client := meta.(*ArmClient).workspacesClient
+	id, err := parseAzureResourceID(d.Id())
+	if err != nil {
+		return err
+	}
+	resGroup := id.ResourceGroup
+	name := id.Path["workspaces"]
+
+	resp, err := client.Get(resGroup, name)
+	if err != nil {
+		if responseWasNotFound(resp.Response) {
+			d.SetId("")
+			return nil
+		}
+		return fmt.Errorf("Error making Read request on AzureRM Operational Insight workspaces '%s': %s", name, err)
+	}
+
+	d.Set("name", resp.Name)
+	d.Set("location", resp.Location)
+	d.Set("resource_group_name", resGroup)
+	d.Set("workspace_id", resp.CustomerID)
+	d.Set("portal_url", resp.PortalURL)
+	d.Set("sku", resp.Sku.Name)
+	d.Set("retention_in_days", resp.RetentionInDays)
+
+	sharedKeys, err := client.GetSharedKeys(resGroup, name)
+	if err != nil {
+		log.Printf("[ERROR] Unable to List Shared keys for Operatinal Insight workspaces %s: %s", name, err)
+	} else {
+		d.Set("primary_shared_key", sharedKeys.PrimarySharedKey)
+		d.Set("secondary_shared_key", sharedKeys.SecondarySharedKey)
+	}
+
+	flattenAndSetTags(d, resp.Tags)
+	return nil
+}
+
+func resourceArmOperationalInsightWorkspaceDelete(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*ArmClient).workspacesClient
+
+	id, err := parseAzureResourceID(d.Id())
+	if err != nil {
+		return err
+	}
+	resGroup := id.ResourceGroup
+	name := id.Path["workspaces"]
+
+	resp, err := client.Delete(resGroup, name)
+
+	if err != nil {
+		if resp.StatusCode == http.StatusNotFound {
+			return nil
+		}
+
+		return fmt.Errorf("Error issuing AzureRM delete request for Operational Insight Workspaces '%s': %+v", name, err)
+	}
+
+	return nil
+}
+
+func getSku(skuName interface{}) (*operationalinsights.Sku, error) {
+	if skuName == nil {
+		return nil, nil
+	}
+	skuEnum, err := getSkuNameEnum(skuName.(string))
+	if err != nil {
+		return nil, err
+	}
+	return &operationalinsights.Sku{
+		Name: skuEnum,
+	}, nil
+}
+
+func getSkuNameEnum(skuName string) (operationalinsights.SkuNameEnum, error) {
+	switch skuName {
+	case "Free":
+		return operationalinsights.Free, nil
+	case "PerNode":
+		return operationalinsights.PerNode, nil
+	case "Premium":
+		return operationalinsights.Premium, nil
+	case "Standalone":
+		return operationalinsights.Standalone, nil
+	case "Standard":
+		return operationalinsights.Standard, nil
+	case "Unlimited":
+		return operationalinsights.Unlimited, nil
+	default:
+		return operationalinsights.Free, fmt.Errorf("Sku name not found")
+	}
+}
+
+func validateAzureRmOperationalInsightWorkspaceName(v interface{}, k string) (ws []string, errors []error) {
+	value := v.(string)
+
+	r, _ := regexp.Compile("^[A-Za-z0-9][A-Za-z0-9-]+[A-Za-z0-9]$")
+	if !r.MatchString(value) {
```

Comment:
I can't see a max length listed in the API Documentation - is there one in the Portal we should add checks for?

Reply:
Name requirements are shown in the azure portal as a tooltip. Workspace name should include 4-63 letters, digits or '-'. The '-' shouldn't be the first or the last symbol.

Reply:
Thank you @piotrgo ! I added the validation with test.